### PR TITLE
EcOp Builtin

### LINKED ISSUE #87: Ec_op builtin

### PR DESCRIPTION
**Pull request by mmsc2**

### CODE REVIEW

File: pkg/builtins/ec_op.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+const INPUT_CELLS_PER_EC_OP = 5
+const CELLS_PER_EC_OP = 7
+
+const PRIME = "0x800000000000011000000000000000000000000000000000000000000000001"
+
+func NewEcOpBuiltinRunner() *EcOpBuiltinRunner {
+	return &EcOpBuiltinRunner{
+		cells_per_instance: CELLS_PER_EC_OP,
+		n_input_cells:      INPUT_CELLS_PER_EC_OP,
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: We can remove these fields and use the constants instead

File: pkg/builtins/ec_op.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+func (ec *EcOpBuiltinRunner) AddValidationRule(*memory.Memory) {}
+
+func (ec *EcOpBuiltinRunner) Base() memory.Relocatable {
+	return ec.base
+}
+
+func (ec *EcOpBuiltinRunner) Name() string {
+	return "ec_op"
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Can we add a constant instead?

Comment (mmsc2) [reply pkg/builtins/ec_op.go]: sure

File: pkg/builtins/ec_op.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	if index != OUTPUT_INDICES.x && index != OUTPUT_INDICES.y {
+		return nil, nil
+	}
+
+	instance := memory.NewRelocatable(address.SegmentIndex, address.Offset-index)
+	input_cells_per_ec_op := lambdaworks.FeltFromUint64(INPUT_CELLS_PER_EC_OP)
+	x_addr, err := instance.AddFelt(input_cells_per_ec_op)
+	if err != nil {
+		return nil, errors.New("Runner error, Expected Integer")
+	}
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]:
```suggestion
	x_addr, err := instance.AddFelt(input_cells_per_ec_op)
	if err != nil {
		return nil, err
	}
```

File: pkg/builtins/ec_op.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+	number := ec.cache[address]
+
+	if !reflect.DeepEqual(number, lambdaworks.Felt{}) {
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go:106]:
```suggestion
	if number.isZero(){
```

File: pkg/builtins/ec_op.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+		return memory.NewMaybeRelocatableFelt(number), nil
+	}
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Zero is aso a valid cache value
```suggestion
	number, is_cached := ec.cache[address]

	if is_cached {
		return memory.NewMaybeRelocatableFelt(number), nil
	}
```

File: pkg/builtins/ec_op.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+	//All input cells should be filled, and be integer values
+	//If an input cell is not filled, return None
+
+	input_cells := make([]lambdaworks.Felt, 0)
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go:118]:
```suggestion
	input_cells := make([]lambdaworks.Felt, INPUT_CELLS_PER_EC_OP)
```
We already know the size of the array, so we can avoid resizing with append

File: pkg/builtins/ec_op.go [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+	for i := 0; i < int(ec.n_input_cells); i++ {
+		addr, err := segments.Get(instance.AddUint(uint(i)))
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go:115]: its not an addr

Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Its not an addr

File: pkg/builtins/ec_op.go [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+	} else {
+		return []memory.MaybeRelocatable{}
+	}
+}
+
+func (ec *EcOpBuiltinRunner) Include(include bool) {
+	ec.included = include
+}
+
+func (ec *EcOpBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, segments *memory.Memory) (*memory.MaybeRelocatable, error) {
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]:
```suggestion
func (ec *EcOpBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, mem *memory.Memory) (*memory.MaybeRelocatable, error) {
```

File: pkg/builtins/ec_op.go [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+	alpha_big_int := big.NewInt(1)
+
+	partial_sum := PartialSum{X: input_cells[0], Y: input_cells[1]}
+	double_point := DoublePoint{X: input_cells[2], Y: input_cells[3]}
+
+	result, err := EcOnImpl(partial_sum, double_point, input_cells[4], alpha_big_int, prime, ec.scalar_height)
+
+	felt_result_x := lambdaworks.FeltFromHex(result.X.Text(16))
+	felt_result_y := lambdaworks.FeltFromHex(result.Y.Text(16))
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Lets use bytes instead of strings for type conversions

File: pkg/builtins/ec_op.go [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+		return memory.NewMaybeRelocatableFelt(felt_result_y), nil
+	}
+}
+
+func LineSlope(point_a PartialSumB, point_b DoublePointB, prime big.Int) (big.Int, error) {
+	mod_value := new(big.Int).Sub(&point_a.X, &point_b.X)
+	mod_value.Mod(mod_value, &prime)
+
+	if mod_value.Cmp(big.NewInt(0)) == 0 {
+		return big.Int{}, errors.New("is multiple of prime")
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Can we make this error more expressive?

File: pkg/builtins/ec_op.go [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+	}
+
+	n := new(big.Int).Sub(&point_a.Y, &point_b.Y)
+	m := new(big.Int).Sub(&point_a.X, &point_b.X)
+
+	z, _ := math_utils.DivMod(n, m, &prime)
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Dont ignore the error

File: pkg/builtins/ec_op.go [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+	y.Sub(y, &point_a.Y)
+	y.Mod(y, &prime)
+
+	return PartialSumB{X: *x, Y: *y}, nil
+}
+
+func EcDoubleSlope(point DoublePointB, alpha big.Int, prime big.Int) (big.Int, error) {
+	q := new(big.Int).Mod(&point.Y, &prime)
+	if q == big.NewInt(0) {
+		return big.Int{}, errors.New("is multiple of prime")
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Can we make this error more expressive?

File: pkg/builtins/ec_op.go [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+	}
+	n := new(big.Int).Mul(&point.X, &point.X)
+	n.Mul(n, big.NewInt(3))
+	n.Add(n, &alpha)
+
+	m := new(big.Int).Mul(&point.Y, big.NewInt(2))
+	z, _ := math_utils.DivMod(n, m, &prime)
```
Comment (fmoletta) [inline pkg/builtins/ec_op.go]: Don't ignore the error

File: pkg/builtins/ec_op.go (hunk 6 above)
Comment (mmsc2) [reply pkg/builtins/ec_op.go:118]: i got a bug that when declaring an array when a fixed size, it fills it with 0s then when i append then it resizes  anyway leaving the 0s

File: pkg/lambdaworks/lib/lambdaworks.h [hunk 14]
```diff
@@ -64,9 +64,18 @@ void felt_or(felt_t a, felt_t b, felt_t result);
...
+
 /* returns the representation of a felt to string */
 char* to_signed_felt(felt_t value);
 
 /* frees a pointer to a string */
 void free_string(char* ptr);
+
+/* writes the result variable with a >> num */
 void felt_shr(felt_t a, size_t b, felt_t result);
+
```
Comment (fmoletta) [inline pkg/lambdaworks/lib/lambdaworks.h]:
```suggestion
```

**Review by jrchatruc — APPROVED** [review]

File: pkg/builtins/ec_op.go (hunk 6 above)
Comment (fmoletta) [reply pkg/builtins/ec_op.go:118]: oh, then we can use `input_cells := make([]lambdaworks.Felt, 0, INPUT_CELLS_PER_EC_OP)` to set the len to 0 and capacity to INPUT_CELLS_PER_EC_OP

**Review by fmoletta — APPROVED** [review]

**Pull request merged**